website/blog/deployers.md

```diff
@@ -0,0 +1,525 @@
+descriptors 3 and 4 (see [the exec package][exec] for details). The weavelet
+running in the subprocess opens the file descriptors stored in the
+`ENVELOPE_TO_WEAVELET_FD` and `WEAVELET_TO_ENVELOPE_FD` environment variables to
+establish its connection to the deployer.
+
+```go
+func run(ctx context.Context, binary string) error {
+    // Step 1...
+
+    // Step 2. Send a WeaveletInfo to the weavelet.
+    info := &protos.EnvelopeMsg{
+        WeaveletInfo: &protos.WeaveletInfo{
+            App:           "app",
+            DeploymentId:  uuid.New().String(),
+            Id:            uuid.New().String(),
+            SingleProcess: false,
+            SingleMachine: true,
+        },
+    }
+    if err := protomsg.Write(deployerWriter, info); err != nil {
+        return err
+    }
+
+    // ...
+}
+```
+
+Second, the deployer sends an `EnvelopeMsg` containing a `WeaveletInfo` to the
+weavelet. This is the first message a weavelet expects to receive. A
+`WeaveletInfo` provides the weavelet with a basic set of metadata including the
+name of the app, the unique id of the deployment, the unique id of the weavelet,
+and so on.
+
+```go
+func run(ctx context.Context, binary string) error {
+    // Step 1...
+    // Step 2...
+
+    // Step 3. Receive and acknowledge a registration from the weavelet.
+    var reply protos.WeaveletMsg
+    if err := protomsg.Read(deployerReader, &reply); err != nil {
+        return err
+    }
+    fmt.Println(protojson.Format(&reply))
+    return nil
+}
+```
+
+Third, the deployer reads and prints a `WeaveletMsg` from the weavelet before
+terminating. After the weavelet receives the `WeaveletInfo` in Step 2, it
+responds with a [ReplicaToRegister][ReplicaToRegister] protobuf with information
+about itself (e.g., its address). Here's an example of what the `WeaveletMsg`
+looks like when printed by our deployer:
+
+```json
+{
+    "id": "1",
+    "replicaToRegister": {
+        "address":      "tcp://100.116.161.153:35077",
+    }
+}
+```
+
+This trivial deployer demonstrates how a deployer and weavelet exchange
+information via protobufs over a pair of pipes. Understanding these low-level
+details is important if you want to implement a deployer completely
+scratch, say in another programming language. For even more details, refer to
+[runtime.proto][].
+
+## Envelopes
+
+Service Weaver provides an `Envelope` API that abstracts away all the tedious
+code we wrote by hand in the previous section. An `Envelope` runs a Service
+Weaver binary as a subprocess; establishes pipes to the weavelet running in the
+subprocess; and serializes and deserializes protos over these pipes.
+
+We construct an `Envelope` with an instance of the [EnvelopeHandler][]
+interface. This interface defines how a deployer handles requests from the
+weavelet.
+
+```go
+type EnvelopeHandler interface {
+    // Responsibility 1: Telemetry.
+    RecvLogEntry(*protos.LogEntry)
```

Comment:
Is it just happenstance that currently RecvsLogEntry() cannot generate errors? For symmetry maybe we should return an error anyway?